server/hotrod/src/main/java/org/infinispan/server/hotrod/ContextHandler.java

```diff
@@ -0,0 +1,204 @@
+            } else {
+               marshaller = new GenericJBossMarshaller();
+            }
+            byte[] result = (byte[]) taskManager.runTask(execContext.name(),
+                    new TaskContext().marshaller(marshaller).cache(msg.cache()).parameters(execContext.params())).get();
+            writeResponse(msg, ctx.channel(),
+                    new ExecResponse(h.version(), h.messageId(), h.cacheName(), h.clientIntel(), h.topologyId(), result));
+            break;
+         case BulkGetRequest:
+            int size = (int) msg.operationDecodeContext();
+            if (msg.isTrace()) {
+               log.tracef("About to create bulk response count = %d", size);
+            }
+            writeResponse(msg, ctx.channel(), new BulkGetResponse(h.version(), h.messageId(), h.cacheName(), h.clientIntel(),
+                    h.topologyId(), size));
+            break;
+         case BulkGetKeysRequest:
+            int scope = (int) msg.operationDecodeContext();
+            if (msg.isTrace()) {
+               log.tracef("About to create bulk get keys response scope = %d", scope);
+            }
+            writeResponse(msg, ctx.channel(), new BulkGetKeysResponse(h.version(), h.messageId(), h.cacheName(), h.clientIntel(),
+                    h.topologyId(), scope));
+            break;
+         case QueryRequest:
+            byte[] queryResult = server.query(msg.cache(), (byte[]) msg.operationDecodeContext());
+            writeResponse(msg, ctx.channel(),
+                    new QueryResponse(h.version(), h.messageId(), h.cacheName(), h.clientIntel(), h.topologyId(), queryResult));
+            break;
+         case AddClientListenerRequest:
+            ClientListenerRequestContext clientContext = (ClientListenerRequestContext) msg.operationDecodeContext();
+            server.getClientListenerRegistry().addClientListener(msg.decoder(), ctx.channel(), h, clientContext.listenerId(),
+                    msg.cache(), clientContext.includeCurrentState(), new Tuple2<>(clientContext.filterFactoryInfo(),
+                            clientContext.converterFactoryInfo()), clientContext.useRawData());
+            break;
+         case RemoveClientListenerRequest:
+            byte[] listenerId = (byte[]) msg.operationDecodeContext();
+            if (server.getClientListenerRegistry().removeClientListener(listenerId, msg.cache())) {
+               writeResponse(msg, ctx.channel(), msg.decoder().createSuccessResponse(h, null));
+            } else {
+               writeResponse(msg, ctx.channel(), msg.decoder().createNotExecutedResponse(h, null));
+            }
+            break;
+         case IterationStartRequest:
+            Tuple4<Option<byte[]>, Option<Tuple2<String, scala.collection.immutable.List<byte[]>>>, Integer, Boolean> iterationStart =
+                    (Tuple4<Option<byte[]>, Option<Tuple2<String, scala.collection.immutable.List<byte[]>>>, Integer, Boolean>) msg.operationDecodeContext();
+
```

Comment:
I think after this PR is integrated we should consider gradually move the scala sources to java, otherwise we end up with those ugly casts everywhere.